### PR TITLE
Add source reference information for merged packages

### LINKED ISSUE #101: Not using revisions
When I add a revision for a project in a "merged" composer.json, composer does not respect the commiit, like:

    "require": {
        "drupal/config_devel": "dev-8.x-1.x#d4cdd21"
    },

Instead of using revision `d4cdd21` it using the latest from HEAD. With a "normal" composer.json (not merged) it works and uses correct revision.

### PR DESCRIPTION
**Pull request by webflo**
Attempt to fix #101

### CODE REVIEW

**Comment by mikkeschiren** [comment]
Real good work @webflo, seems to work very well, this is my composer.json:

````
{
    "name": "nodeone/wkse",
    "description": "Installation profile for wkse",
    "type": "drupal-profile",
    "minimum-stability": "dev",
    "repositories": [
      {
      "type": "package",
        "package":
          {
            "name": "md-systems/pathauto",
            "type": "drupal-module",
            "version": "8.1.0",
              "source": {
                "url": "https://github.com/md-systems/pathauto.git",
                "type": "git",
                "reference": "a38509d"
              }
          }
      },
      {
        "type": "composer",
        "url": "https://packagist.drupal-composer.org"
      }
    ],
    "require": {
        "drupal/devel": "dev-8.x-1.x#d913e1d",
        "drupal/config_devel": "dev-8.x-1.x#d4cdd21",
        "md-systems/pathauto": "^8.1",
        "drupal/token": "dev-8.x-1.x#d6cd94f19b053d8b6a411cb8fc5e75112d141d77",
        "drupal/memcache": "8.2.*@dev"
    }
`````

And this was the result:

`````
Updating dependencies (including require-dev)
  - Installing md-systems/pathauto (8.1.0)
    Cloning a38509d

  - Installing drupal/devel (dev-8.x-1.x d913e1d)
    Cloning d913e1d

  - Installing drupal/config_devel (dev-8.x-1.x d4cdd21)
    Cloning d4cdd21

  - Installing drupal/token (dev-8.x-1.x d6cd94f)
    Cloning d6cd94f19b053d8b6a411cb8fc5e75112d141d77

  - Installing drupal/memcache (dev-8.x-2.x 222f62f)
    Cloning 222f62f7c462f084d7d1852f747e0bf7e16909ca
`````

File: src/Merge/ExtraPackage.php [hunk 1]
```diff
@@ -165,6 +171,20 @@ public function mergeInto(RootPackageInterface $root, PluginState $state)
...
+        foreach (array('require', 'require-dev') as $linkType) {
+            $linkInfo = BasePackage::$supportedLinkTypes[$linkType];
+            $method = 'get'.ucfirst($linkInfo['method']);
+            $links = array();
+            foreach ($root->$method() as $link) {
+                $links[$link->getTarget()] = $link->getConstraint()->getPrettyString();
+            }
+            $references = $this->extractReferences($links, $references);
+        }
+        $root->setReferences($references);
```
Comment (webflo) [inline src/Merge/ExtraPackage.php]: @bd808 setReferences is not part of the ``RootPackageInteface``. Should we use ``unwrapIfNeeded`` or should we try to fix it upstream?

Comment (bd808) [inline src/Merge/ExtraPackage.php]: I'd say that both are wanted. At the very least `unwrapIfNeeded` will cast or provide access to the aliased `RootPackage`. Upstreaming interface additions and tests is always a good idea I think. See https://github.com/composer/composer/commit/06c44ce998cdb0db2a522ec6086d6ce25bd79b4a for an example of us doing this previously.

**Comment by bd808** [comment]
Commits should be squashed and rebased on current master.

**Comment by LionsAd** [comment]
FWIW, The patch looks great to me!

**Pull request merged**